car-test.mjs

```diff
@@ -100,10 +100,10 @@ async function run () {
       console.log('Got cidC', blockC, blockC === undefined ? 'WRONG!!' : '')
     }
 
-    read()
     if (!skipWrite) {
-      write()
+      await write()
     }
+    await read()
```

Comment:
this section is the biggest problem, the `read()` is giving a 1000ms delay before diving in and decoding the CAR that's already there, but only part of it is there since the write hasn't finished. What's more, the plain `CarReader.fromX()` is going to do a _full_ read and decode into memory of a CAR, no updating, just slurps it all in. So because we're not waiting to read the CAR before it's finished, you're only parsing a partial CAR, hence the keys not existing, because they didn't exist at the time of reading.

Reply:
I wrote it that way to try to reproduce the behaviour I was observing with web3.storage and ipfs-car ... I'm unable to store my dataset because it has zero length files in it. This is what is happening with that code.

If it's a requirement that the CAR has to be fully written out before it can be read, that's good information and I can submit a PR against the web3.storage client I think.